### ISSUE TITLE
Optimize equivariant transformer

### ISSUE THREAD

**Issue by peastman** [torchmd/torchmd-net#45]
I've started looking into how to improve the speed of the equivariant transformer.  I'm opening this issue as a place to report my progress.

I'm mainly interested in the speed of running simulations with it: predicting energy and forces for a single molecule.  I'm less concerned about training speed.  As a starting point, I tried loading molecules of various sizes and computing forces.  For the small molecules we're mainly interested in, the time is completely dominated by overhead and barely affected by the number of atoms.  It ranged from about 11.5 ms for molecules with around 20 atoms, up to 13.1 ms for one with 94 atoms.

Viewing it with the PyTorch profiler shows the GPU running lots of tiny kernels with long gaps in between.  The GPU is sitting idle most of the time.  In contrast, if I profile it during training with a batch size of 64, the GPU usage is very high.  There might still be ways to make it faster, but at least operating on batches keeps the GPU from being idle.

There are two possible approaches to making it faster: improve the PyTorch code, or just replace it with custom CUDA kernels.  The latter is likely to be more effective, so that's what I'm inclined to do.  The custom kernels would be added to NNPOps.

**Comment by peastman** [comment]
Any thoughts on this question?  Is there actually a reason to change the functional form, or is it really just a matter of not putting RBFs at short distances?

I found that the lower cutoff also changes the model in a significant, unrelated way.  When accumulating interactions, you multiply the attention weight by the cutoff function.

https://github.com/torchmd/torchmd-net/blob/57b4d51f222f3f66405f6b4c71511099144ebd19/torchmdnet/models/torchmd_et.py#L308

Without a lower cutoff, the cutoff function goes to 1 at short distances, which allows self interactions.  You get contributions to the sum for each atom interacting with itself.  With a lower cutoff, the cutoff function goes to 0 which eliminates self interactions.  Did you really intend it to have that effect?  If not, which behavior did you intend (with or without self interactions)?

**Comment by PhilippThoelke** [comment]
We actually start distributing the RBFs only beyond the lower cutoff in `GaussianSmearing` and `ExpNormalSmearing`. This does not require changing the functional form and in fact we don't even use the cutoff function with a lower cutoff in these two classes.

The change in functional form is important where we want to discard interactions between atoms below the lower cutoff as in the message passing step of the graph network:
https://github.com/torchmd/torchmd-net/blob/57b4d51f222f3f66405f6b4c71511099144ebd19/torchmdnet/models/torchmd_gn.py#L247-L248
Here it is not enough to rely on the RBF distribution as the filter generating network (`self.net`) might produce non-zero entries in the filter even if all inputs are 0.

Regarding the ET, however, it is not intended that self interactions disappear when lower cutoff is larger than 0. The model should definitely have self interactions. Maybe we could explicitly allow interactions if the distance `r_ij == 0`.

**Comment by PhilippThoelke** [comment]
We do not want to throw away self interactions. I believe the locations where this is currently happening with lower cutoffs larger 0 are your 1st, 5th and 6th points. However, it is still important that we restrict the attention mechanism to interactions between the lower and upper cutoff inside `EquivariantMultiHeadAttention` if we want to keep the lower cutoff as an option.

I personally haven't used the lower cutoff so far but as far as I know it has helped in other projects, however, using the graph network and not the ET. Maybe it makes sense to remove the lower cutoff option for the ET (or altogether)? @giadefa 

> The correct solution, it seems to me, is just to make sure your training data includes close contacts so the model can accurately learn that they produce huge energies.

I completely agree but depending on the type of application it might be very difficult to obtain data, which includes close contacts (e.g. existing MD data, which is too expensive to rebuild/augment with close contacts).

> When two atoms come very close together, that has a huge effect on the energy. Forcing it to instead have no effect at all won't produce anything the least bit realistic.

That is if the neural network potential is the only component in the energy calculation. Poorly sampled regions, like close contacts in MD, can be handled by a more robust but less accurate prior model.

**Comment by giadefa** [comment]
So, when two beads come together the NN energy is just anywhere and so the
forces. Forces can be strongly attractive at 1A and strongly repulsive at
1.2A. Ideally one would like to regularize the forces, but this involves
the gradient of the gradient of the energy in the loss.
A lower cutoff allows for zero forces in that region and it is the only
solution we found for coarse-graining simulations.

For quantum data the lower cutoff is less important, so you might just
ignore it, if it is simpler. We only use it  for coarse-graining.

g

**Comment by giadefa** [comment]
For me it is fine to remove the lower cutoff for ET

**Comment by peastman** [comment]
> Poorly sampled regions, like close contacts in MD, can be handled by a more robust but less accurate prior model.

Perhaps a way of handling that would be to add a regularization term encouraging the NN energy contribution to go to zero at short distances?  Or just create some artificial training samples that represent very close contacts and specify the energy as equal to the energy of the prior model?

**Comment by giadefa** [comment]
We tried with the training samples at low distance but did not work,
because the NN is many-body. You can fix the two-body but then you have
problems with 3 body interactions going wild.
